### What is EDA with Python?
Exploratory Data Analysis (EDA) with Python involves analyzing and summarizing data to gain insights and understand its underlying patterns, relationships, and distributions using Python programming language.

### How to make EDA in Python?
To perform EDA in Python, you can use libraries like Pandas, NumPy, Matplotlib, and Seaborn. These libraries provide functions and tools for data manipulation, visualization, and statistical analysis, which facilitate the process of exploring and understanding the data

### How to perform EDA in machine learning?
Performing EDA in machine learning typically involves preprocessing the data by handling missing values, outliers, and feature scaling. Then, various statistical and visual techniques can be employed to analyze the relationships between variables, identify patterns, and assess the relevance of features. This helps in gaining a better understanding of the data before building a machine learning model

### Complete process to do EDA?
- Step 1: Import Python Libraries
- Step 2: Reading Dataset
- Step 3: Data Reduction 
- Step 4: Creating Features
- Step 5: Data Cleaning/Wrangling
- Step 6: EDA Exploratory Data Analysis
- Step 7: Statistics Summary
- Step 8: EDA Univariate Analysis
- Step 9: Data Transformation
- Step 10: EDA Bivariate Analysis
- Step 11: EDA Multivariate Analysis
- Step 12: Impute Missing values
- Step 13: Feature Engineering
- Conclusion

### Download dataset from here
link: https://www.kaggle.com/datasets/sukhmanibedi/cars4u

## Step 1: Import Python Libraries 

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

C:\Users\Sarthak Aggarwal\AppData\Roaming\Python\Python39\site-packages\pandas\core\computation\expressions.py:21: UserWarning: Pandas requires version '2.8.4' or newer of 'numexpr' (version '2.8.3' currently installed).
  from pandas.core.computation.check import NUMEXPR_INSTALLED
C:\Users\Sarthak Aggarwal\AppData\Roaming\Python\Python39\site-packages\pandas\core\arrays\masked.py:60: UserWarning: Pandas requires version '1.3.6' or newer of 'bottleneck' (version '1.3.5' currently installed).
  from pandas.core import (


## Step 2: Reading Dataset

In [2]:
df=pd.read_csv("cars.csv")

In [3]:
df.head()
# display top 5 values of dataset

S.No.                              Name    Location  Year  \
0      0            Maruti Wagon R LXI CNG      Mumbai  2010   
1      1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
2      2                      Honda Jazz V     Chennai  2011   
3      3                 Maruti Ertiga VDI     Chennai  2012   
4      4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              72000       CNG       Manual      First  26.6 km/kg   998 CC   
1              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
2              46000    Petrol       Manual      First   18.2 kmpl  1199 CC   
3              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
4              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   

       Power  Seats  New_Price  Price  
0  58.16 bhp    5.0        NaN   1.75  
1  126.2 bhp    5.0        NaN  12.50  
2   88.7 bhp    5.0  8.61 Lakh   4.50  
3  88.76 bhp    7.0        NaN   6.00  
4  140.8 bhp    5.0        NaN  17.74

In [4]:
df.columns

Index(['S.No.', 'Name', 'Location', 'Year', 'Kilometers_Driven', 'Fuel_Type',
       'Transmission', 'Owner_Type', 'Mileage', 'Engine', 'Power', 'Seats',
       'New_Price', 'Price'],
      dtype='object')

#### About DataSet:
1. S.No. : Serial Number
2. Name : Name of the car which includes Brand name and Model name
3. Location : The location in which the car is being sold or is available for purchase Citiesr> 4. Year : Manufacturing year of the car
5. Kilometers_driven : The total kilometers driven in the car by the previous owner(s) in KM.
6. Fuel_Type : The type of fuel used by the car. (Petrol, Diesel, Electric, CNG, LPG)
7. Transmission : The type of transmission used by the car. (Automatic / Manual)
8. Owner : Type of ownership
9. Mileage : The standard mileage offered by the car company in kmpl or km/kg
10. Engine : The displacement volume of the engine in CC.
11. Power : The maximum power of the engine in bhp.
12. Seats : The number of seats in the car.
13. New_Price : The price of a new car of the same model in INR Lakhs.(1 Lakh = 100, 000)
14. Price : The price of the used car in INR Lakhs (1 Lakh = 100, 000)

In [5]:
df.shape

(7253, 14)

There are 7253 observations and 14 variables in our dataset

In [6]:
df.info()
#info() helps to understand the data type and information about data, including the number of records in each column, data having null or not null, Data type, the memory usage of the dataset

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7253 entries, 0 to 7252
Data columns (total 14 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   S.No.              7253 non-null   int64  
 1   Name               7253 non-null   object 
 2   Location           7253 non-null   object 
 3   Year               7253 non-null   int64  
 4   Kilometers_Driven  7253 non-null   int64  
 5   Fuel_Type          7253 non-null   object 
 6   Transmission       7253 non-null   object 
 7   Owner_Type         7253 non-null   object 
 8   Mileage            7251 non-null   object 
 9   Engine             7207 non-null   object 
 10  Power              7207 non-null   object 
 11  Seats              7200 non-null   float64
 12  New_Price          1006 non-null   object 
 13  Price              6019 non-null   float64
dtypes: float64(2), int64(3), object(9)
memory usage: 793.4+ KB


It shows the variables Mileage, Engine, Power, Seats, New_Price, and Price have missing values.

In [7]:
# Check for duplication
df.nunique()

# tells about unique values in each column

S.No.                7253
Name                 2041
Location               11
Year                   23
Kilometers_Driven    3660
Fuel_Type               5
Transmission            2
Owner_Type              4
Mileage               450
Engine                150
Power                 386
Seats                   9
New_Price             625
Price                1373
dtype: int64

### Missing value calculation

In [8]:
df.isnull().sum()

S.No.                   0
Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage                 2
Engine                 46
Power                  46
Seats                  53
New_Price            6247
Price                1234
dtype: int64

In [9]:
(df.isnull().sum()/(len(df)))*100

S.No.                 0.000000
Name                  0.000000
Location              0.000000
Year                  0.000000
Kilometers_Driven     0.000000
Fuel_Type             0.000000
Transmission          0.000000
Owner_Type            0.000000
Mileage               0.027575
Engine                0.634220
Power                 0.634220
Seats                 0.730732
New_Price            86.129877
Price                17.013650
dtype: float64

#### Observations:

This preview shows that some columns potentially have a lot of missingness so we'll want to make sure to look into that later.
- New_Price has only 1006 values. 86 % values are missing
- Price, which is a Target variable 17 % missing values.This needs to be analysed further.
- Seats has only 53 values missing and number of seats can be one of key factor in deciding price.
- Power and Engine has 46 missing values.
- Mileage only has two values missing.
- Mileage,Power,Engine,New_Price we know are quantitative variables but are of object dtype here and needs to to converted to numeric.

## Step 3: Data Reduction
Some columns or variables can be dropped if they do not add value to our analysis.

In our dataset, the column S.No have only ID values, assuming they don’t have any predictive power to predict the dependent variable.

In [10]:
df.drop(['S.No.'],axis=1,inplace=True)

In [11]:
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  
0        NaN   1.75  
1        NaN  12.50  
2  8.61 Lakh   4.50  
3        NaN   6.00  
4        NaN  17.74

## Step 4: Creating Features
It would be difficult to find the car’s age if it is in year format as the Age of the car is a contributing factor to Car Price. 

Introducing a new column, “Car_Age” to know the age of the car

In [12]:
from datetime import date
year_now=date.today().year
df['Car_Age']=year_now-df['Year']
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  Car_Age  
0        NaN   1.75       14  
1        NaN  12.50        9  
2  8.61 Lakh   4.50       13  
3        NaN   6.00       12  
4        NaN  17.74       11

- Since car names will not be great predictors of the price in our current data. But we can process this column to extract important information using brand and Model names. Let’s split the name and introduce new variables “Brand” and “Model”

In [13]:
df['Brand'] = df.Name.str.split().str.get(0)
df['Model'] = df.Name.str.split().str.get(1) + df.Name.str.split().str.get(2)
df[['Name','Brand','Model']]

Name          Brand  \
0                                Maruti Wagon R LXI CNG         Maruti   
1                      Hyundai Creta 1.6 CRDi SX Option        Hyundai   
2                                          Honda Jazz V          Honda   
3                                     Maruti Ertiga VDI         Maruti   
4                       Audi A4 New 2.0 TDI Multitronic           Audi   
...                                                 ...            ...   
7248                  Volkswagen Vento Diesel Trendline     Volkswagen   
7249                             Volkswagen Polo GT TSI     Volkswagen   
7250                             Nissan Micra Diesel XV         Nissan   
7251                             Volkswagen Polo GT TSI     Volkswagen   
7252  Mercedes-Benz E-Class 2009-2013 E 220 CDI Avan...  Mercedes-Benz   

                 Model  
0               WagonR  
1             Creta1.6  
2                JazzV  
3            ErtigaVDI  
4                A4New  
...                ...  
7248       VentoDiesel  
7249            PoloGT  
7250       MicraDiesel  
7251            PoloGT  
7252  E-Class2009-2013  

[7253 rows x 3 columns]

In [14]:
df.head()

Name    Location  Year  Kilometers_Driven  \
0            Maruti Wagon R LXI CNG      Mumbai  2010              72000   
1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
2                      Honda Jazz V     Chennai  2011              46000   
3                 Maruti Ertiga VDI     Chennai  2012              87000   
4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0       CNG       Manual      First  26.6 km/kg   998 CC  58.16 bhp    5.0   
1    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
2    Petrol       Manual      First   18.2 kmpl  1199 CC   88.7 bhp    5.0   
3    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
4    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   

   New_Price  Price  Car_Age    Brand      Model  
0        NaN   1.75       14   Maruti     WagonR  
1        NaN  12.50        9  Hyundai   Creta1.6  
2  8.61 Lakh   4.50       13    Honda      JazzV  
3        NaN   6.00       12   Maruti  ErtigaVDI  
4        NaN  17.74       11     Audi      A4New

In [15]:
df['Power_num'] = df.Power.str.split().str.get(0)
df['Power_Unit'] = df.Power.str.split().str.get(1) 
y=df[['Power','Power_num','Power_Unit']]

In [16]:
y

Power Power_num Power_Unit
0     58.16 bhp     58.16        bhp
1     126.2 bhp     126.2        bhp
2      88.7 bhp      88.7        bhp
3     88.76 bhp     88.76        bhp
4     140.8 bhp     140.8        bhp
...         ...       ...        ...
7248  103.6 bhp     103.6        bhp
7249  103.6 bhp     103.6        bhp
7250   63.1 bhp      63.1        bhp
7251  103.6 bhp     103.6        bhp
7252    170 bhp       170        bhp

[7253 rows x 3 columns]

In [17]:
y.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7253 entries, 0 to 7252
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Power       7207 non-null   object
 1   Power_num   7207 non-null   object
 2   Power_Unit  7207 non-null   object
dtypes: object(3)
memory usage: 170.1+ KB


In [18]:
df['Power_num'] = pd.to_numeric(df['Power_num'])

ValueError: Unable to parse string "null" at position 76

In [23]:
df = df[df['Power_num'] != 'null']

In [24]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 7124 entries, 0 to 7252
Data columns (total 18 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               7124 non-null   object 
 1   Location           7124 non-null   object 
 2   Year               7124 non-null   int64  
 3   Kilometers_Driven  7124 non-null   int64  
 4   Fuel_Type          7124 non-null   object 
 5   Transmission       7124 non-null   object 
 6   Owner_Type         7124 non-null   object 
 7   Mileage            7122 non-null   object 
 8   Engine             7078 non-null   object 
 9   Power              7078 non-null   object 
 10  Seats              7075 non-null   float64
 11  New_Price          1006 non-null   object 
 12  Price              5912 non-null   float64
 13  Car_Age            7124 non-null   int64  
 14  Brand              7124 non-null   object 
 15  Model              7123 non-null   object 
 16  Power_num          7078 non-n

In [25]:
df.nunique()

Name                 1991
Location               11
Year                   22
Kilometers_Driven    3610
Fuel_Type               5
Transmission            2
Owner_Type              4
Mileage               448
Engine                144
Power                 385
Seats                   8
New_Price             625
Price                1368
Car_Age                22
Brand                  31
Model                 705
Power_num             385
Power_Unit              1
dtype: int64

In [26]:
df["Power_num"].unique

<bound method Series.unique of 0       58.16
1       126.2
2        88.7
3       88.76
4       140.8
        ...  
7248    103.6
7249    103.6
7250     63.1
7251    103.6
7252      170
Name: Power_num, Length: 7124, dtype: object>

In [ ]:
s

## Step 5: Data Cleaning/Wrangling
Some names of the variables are not relevant and not easy to understand. Some data may have data entry errors, and some variables may need data type conversion. We need to fix this issue in the data

In [ ]:
print(df.Brand.unique())
print(df.Brand.nunique())

The brand name ‘Isuzu’ ‘ISUZU’ and ‘Mini’ and ‘Land’ looks incorrect. This needs to be corrected

In [ ]:
searchfor=['Isuzu' ,'ISUZU','Mini','Land']
df[df.Brand.str.contains('|'.join(searchfor))]

In [ ]:
df["Brand"].replace({"ISUZU": "Isuzu", "Mini": "Mini Cooper","Land":"Land Rover"}, inplace=True)

In [ ]:
print(df.Brand.unique())
print(df.Brand.nunique())

The data cleaning process involves removing duplicate or irrelevant observations, fixing structural errors, filtering unwanted outliers, handling missing data, and validating and QA.                      
We will do remaining data cleaning after performing EDA

## Step 6: EDA Exploratory Data Analysis
Exploratory Data Analysis refers to the crucial process of performing initial investigations on data to discover patterns to check assumptions with the help of summary statistics and graphical representations.
- EDA can be leveraged to check for outliers, patterns, and trends in the given data.
- EDA helps to find meaningful patterns in data.
- EDA provides in-depth insights into the data sets to solve our business problems.
- EDA gives a clue to impute missing values in the dataset 

## Step 7: Statistics Summary
The information gives a quick and simple description of the data.

Can include Count, Mean, Standard Deviation, median, mode, minimum value, maximum value, range, standard deviation, etc.
- count - The number of not-empty values.
- mean - The average (mean) value.
- std - The standard deviation.
- min - the minimum value.
- 25% - The 25% percentile*.
- 50% - The 50% percentile*.
- 75% - The 75% percentile*.
- max - the maximum value.

Statistics summary gives a high-level idea to identify whether the data has any outliers, data entry error, distribution of data such as the data is normally distributed or left/right skewed

In [ ]:
df.describe()
#describe()– Provide a statistics summary of data belonging to numerical datatype such as int, float

From the statistics summary, we can infer the below findings :
- Years range from 1996- 2019 and has a high in a range which shows used cars contain both latest models and old model cars.
- On average of Kilometers-driven in Used cars are ~58k KM. The range shows a huge difference between min and max as max values show 650000 KM shows the evidence of an outlier. This record can be removed.
- Min value of Mileage shows 0 cars won’t be sold with 0 mileage. This sounds like a data entry issue.
- It looks like Engine and Power have outliers, and the data is right-skewed.
- The average number of seats in a car is 5. car seat is an important feature in price contribution.
- The max price of a used car is 160k which is quite weird, such a high price for used cars. There may be an outlier or data entry issue.

In [ ]:
# divide columns into categorical and numerical columns
cat_cols=df.select_dtypes(include=['object']).columns
num_cols=df.select_dtypes(include=np.number).columns.tolist()
print("Categorical Variables:")
print(cat_cols)
print("Numerical Variables:")
print(num_cols)

## Step 8: EDA Univariate Analysis
Analyzing/visualizing the dataset by taking one variable at a time:

Data visualization is essential; we must decide what charts to plot to better understand the data. In this article, we visualize our data using Matplotlib and Seaborn libraries.

Univariate analysis can be done for both Categorical and Numerical variables.

Categorical variables can be visualized using a Count plot, Bar Chart, Pie Plot, etc.

Numerical Variables can be visualized using Histogram, Box Plot, Density Plot, etc.

In our example, we have done a Univariate analysis using Histogram and  Box Plot for continuous Variables.

In [ ]:
# Loop through each column in the list of numerical columns
for col in num_cols:
    # Print the name of the current column
    print(col)
    # Print the skewness of the data in the current column, rounded to 2 decimal places
    print('Skew :', round(df[col].skew(), 2))
    
    # Create a new figure for plotting with a specific size
    plt.figure(figsize=(15, 4))
    
    # Create subplots, 1 row and 2 columns, and set the current subplot to the first one
    plt.subplot(1, 2, 1)
    
    # Plot a histogram of the data in the current column without grid lines
    df[col].hist(grid=False)
    
    # Label the y-axis of the histogram
    plt.ylabel('count')
    
    # Set the current subplot to the second one
    plt.subplot(1, 2, 2)
    
    # Plot a boxplot of the data in the current column
    sns.boxplot(x=df[col])
    
    # Display the plot
    plt.show()

Skewness is a measure of the asymmetry of the probability distribution of a real-valued random variable about its mean. In simpler terms, it indicates the degree to which the data is skewed or distorted from a normal distribution.

Skewness is often calculated as a numerical value, where a skewness value of 0 indicates a symmetrical distribution, positive values indicate right skew, and negative values indicate left skew. Skewness is a useful statistic because it provides information about the shape of the distribution, which can be important for understanding the characteristics of the data.

1. Year:
- Skew: -0.84
- Since the skewness is negative, it indicates a left skew in the distribution of the 'Year' variable.
- In this case, it suggests that most cars in your dataset were manufactured in recent years, with fewer cars from earlier years. This is typical in datasets involving items that have a temporal aspect, as newer items are often more prevalent.

2. Kilometers_Driven:
- Skew: 61.58
- This skewness value is extremely high, indicating a heavily right-skewed distribution.
- It suggests that the majority of cars in your dataset have very low values of 'Kilometers_Driven', with only a few cars having extremely high values. This could imply that most cars have relatively low mileage, but there are a few outliers with exceptionally high mileage.

3. Seats:
- Skew: 1.9
- The positive skewness indicates a right skew in the distribution of the 'Seats' variable.
- This suggests that most cars in your dataset have fewer seats, with a few cars having a higher number of seats. It's common for cars to have a standard number of seats, with fewer cars having more seats, leading to this right skew.

4. Price:
- Skew: 3.34
- With a high positive skewness, the distribution of 'Price' is heavily skewed to the right.
- This indicates that most cars in your dataset have relatively low prices, while a few cars have exceptionally high prices. It's typical in markets like the automotive industry to have a few luxury or high-end vehicles with significantly higher prices compared to the majority of cars.

5. Car_Age:
- Skew: 0.84
- This positive skewness suggests a right skew in the distribution of 'Car_Age'.
- It implies that most cars in your dataset are relatively young, with fewer older cars. This is consistent with the expectation that newer cars are more common in the market, with older cars being less prevalent

In [ ]:
# categorical variables are being visualized using a count plot. Categorical variables provide the pattern of factors influencing car price
# Create a grid of subplots with 3 rows and 2 columns, with a specific size
fig, axes = plt.subplots(3, 2, figsize=(18, 18))

# Add a title to the entire figure
fig.suptitle('Bar plot for all categorical variables in the dataset')

# Plot a count plot for the 'Fuel_Type' variable in the first row, first column
sns.countplot(ax=axes[0, 0], x='Fuel_Type', data=df, color='blue', 
              order=df['Fuel_Type'].value_counts().index)

# Plot a count plot for the 'Transmission' variable in the first row, second column
sns.countplot(ax=axes[0, 1], x='Transmission', data=df, color='blue', 
              order=df['Transmission'].value_counts().index)

# Plot a count plot for the 'Owner_Type' variable in the second row, first column
sns.countplot(ax=axes[1, 0], x='Owner_Type', data=df, color='blue', 
              order=df['Owner_Type'].value_counts().index)

# Plot a count plot for the 'Location' variable in the second row, second column
sns.countplot(ax=axes[1, 1], x='Location', data=df, color='blue', 
              order=df['Location'].value_counts().index)

# Plot a count plot for the 'Brand' variable in the third row, first column
sns.countplot(ax=axes[2, 0], x='Brand', data=df, color='blue', 
              order=df['Brand'].head(20).value_counts().index)

# Plot a count plot for the 'Model' variable in the third row, second column
sns.countplot(ax=axes[2, 1], x='Model', data=df, color='blue', 
              order=df['Model'].head(20).value_counts().index)

# Rotate the x-axis labels of the plot in the second row, second column by 45 degrees
axes[1, 1].tick_params(labelrotation=45)

# Rotate the x-axis labels of the plot in the third row, first column by 90 degrees
axes[2, 0].tick_params(labelrotation=90)

# Rotate the x-axis labels of the plot in the third row, second column by 90 degrees
axes[2, 1].tick_params(labelrotation=90)

# Display the plot
plt.show()

From the count plot, we can have below observations:
- Mumbai has the highest number of cars available for purchase, followed by Hyderabad and Coimbatore
- ~53% of cars have fuel type as Diesel this shows diesel cars provide higher performance
- ~72% of cars have manual transmission
- ~82 % of cars are First owned cars. This shows most of the buyers prefer to purchase first-owner cars
- ~20% of cars belong to the brand Maruti followed by 19% of cars belonging to Hyundai
- WagonR ranks first among all models which are available for purchase

## Step 9: Data Transformation
Before we proceed to Bi-variate Analysis, Univariate analysis demonstrated the data pattern as some variables to be transformed.

Price and Kilometer-Driven variables are highly skewed and on a larger scale. Let’s do log transformation.

Log transformation can help in normalization, so this variable can maintain standard scale with other variables.Logarithmic transformations can reduce the impact of outliers by compressing the data's range and bringing extreme values closer to the mean

In [ ]:
# Function for log transformation of the column
def log_transform(data,col):
    for colname in col:
        if (data[colname] == 1.0).all():
            data[colname + '_log'] = np.log(data[colname]+1)
        else:
            data[colname + '_log'] = np.log(data[colname])
    data.info()

In [ ]:
log_transform(df,['Kilometers_Driven','Price'])

In [ ]:
#Log transformation of the feature 'Kilometers_Driven'
sns.distplot(df["Kilometers_Driven_log"], axlabel="Kilometers_Driven_log")

In [ ]:
#Log transformation of the feature 'Kilometers_Driven'
sns.distplot(df["Price_log"], axlabel="Price_log")

## Step 10: EDA Bivariate Analysis
Now, let’s move ahead with bivariate analysis. Bivariate Analysis helps to understand how variables are related to each other and the relationship between dependent and independent variables present in the dataset.

For Numerical variables, Pair plots and Scatter plots are widely been used to do Bivariate Analysis.

A Stacked bar chart can be used for categorical variables if the output variable is a classifier. Bar plots can be used if the output variable is continuous

In our example, a pair plot has been used to show the relationship between two Categorical variables.

In [ ]:
plt.figure(figsize=(13,17))
df1=df.drop(['Kilometers_Driven','Price'],axis=1)
sns.pairplot(df1)
plt.show()

Pair Plot provides below insights:
- The variable Year has a positive correlation with price and mileage
- A year has a Negative correlation with kilometers-Driven
- Mileage is negatively correlated with Power
- As power increases, mileage decreases
- Car with recent make is higher at prices. As the age of the car increases price decreases
- Engine and Power increase, and the price of the car increases

A bar plot can be used to show the relationship between Categorical variables and continuous variables 

In [ ]:
fig, axarr = plt.subplots(4, 2, figsize=(12, 18))
df.groupby('Location')['Price_log'].mean().sort_values(ascending=False).plot.bar(ax=axarr[0][0], fontsize=12)
axarr[0][0].set_title("Location Vs Price", fontsize=18)
df.groupby('Transmission')['Price_log'].mean().sort_values(ascending=False).plot.bar(ax=axarr[0][1], fontsize=12)
axarr[0][1].set_title("Transmission Vs Price", fontsize=18)
df.groupby('Fuel_Type')['Price_log'].mean().sort_values(ascending=False).plot.bar(ax=axarr[1][0], fontsize=12)
axarr[1][0].set_title("Fuel_Type Vs Price", fontsize=18)
df.groupby('Owner_Type')['Price_log'].mean().sort_values(ascending=False).plot.bar(ax=axarr[1][1], fontsize=12)
axarr[1][1].set_title("Owner_Type Vs Price", fontsize=18)
df.groupby('Brand')['Price_log'].mean().sort_values(ascending=False).head(10).plot.bar(ax=axarr[2][0], fontsize=12)
axarr[2][0].set_title("Brand Vs Price", fontsize=18)
df.groupby('Model')['Price_log'].mean().sort_values(ascending=False).head(10).plot.bar(ax=axarr[2][1], fontsize=12)
axarr[2][1].set_title("Model Vs Price", fontsize=18)
df.groupby('Seats')['Price_log'].mean().sort_values(ascending=False).plot.bar(ax=axarr[3][0], fontsize=12)
axarr[3][0].set_title("Seats Vs Price", fontsize=18)
df.groupby('Car_Age')['Price_log'].mean().sort_values(ascending=False).plot.bar(ax=axarr[3][1], fontsize=12)
axarr[3][1].set_title("Car_Age Vs Price", fontsize=18)
plt.subplots_adjust(hspace=1.0)
plt.subplots_adjust(wspace=.5)
sns.despine()
plt.show()

#### Observations:
- The price of cars is high in Coimbatore and less price in Kolkata and Jaipur
- Automatic cars have more price than manual cars.
- Diesel and Electric cars have almost the same price, which is maximum, and LPG cars have the lowest price
- First-owner cars are higher in price, followed by a second
- The third owner’s price is lesser than the Fourth and above
- Lamborghini brand is the highest in price
- Gallardocoupe Model is the highest in price
- 2 Seater has the highest price followed by 7 Seater
- The latest model cars are high in price

## Step 11: Multivariate Analysis
As the name suggests, Multivariate analysis looks at more than two variables. Multivariate analysis is one of the most useful methods to determine relationships and analyze patterns for any dataset.

A heat map is widely been used for Multivariate Analysis

Heat Map gives the correlation between the variables, whether it has a positive or negative correlation.

In [ ]:
df2 = df.select_dtypes(exclude=['object'])
# taking only numerical columns

In [ ]:
plt.figure(figsize=(12, 7))
sns.heatmap(df2.corr(), annot = True, vmin = -1, vmax = 1)
plt.show()

#### Observations:
- Price is moderately positive in correlation to year.
- Kilometer driven has a negative correlation to year not much impact on the price
- Car age has a negative correlation with Price
- car Age is positively correlated to Kilometers-Driven as the Age of the car increases

## Step 14: Impute Missing values
Missing data arise in almost all statistical analyses. There are many ways to impute missing values; we can impute the missing values by their Mean, median, most frequent, or zero values and use advanced imputation algorithms like KNN, Regularization, etc.

## Step 13: Feature Engineering
Feature engineering refers to the process of using domain knowledge to select and transform the most relevant variables from raw data when creating a predictive model using machine learning or statistical modeling. The main goal of Feature engineering is to create meaningful data from raw data.